Machine.c

```diff
@@ -0,0 +1,60 @@
+/*
+htop - Machine.c
+(C) 2023 Red Hat, Inc.
+(C) 2004,2005 Hisham H. Muhammad
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "Machine.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "CRT.h"
+#include "Hashtable.h"
+#include "Macros.h"
+#include "Platform.h"
+#include "XUtils.h"
+
+
+void Machine_init(Machine* this, UsersTable* usersTable, uid_t userId) {
+   this->usersTable = usersTable;
+   this->userId = userId;
+
+   // always maintain valid realtime timestamps
+   Platform_gettime_realtime(&this->realtime, &this->realtimeMs);
+
+#ifdef HAVE_LIBHWLOC
+   this->topologyOk = false;
+   if (hwloc_topology_init(&this->topology) == 0) {
+      this->topologyOk =
+         #if HWLOC_API_VERSION < 0x00020000
+         /* try to ignore the top-level machine object type */
+         0 == hwloc_topology_ignore_type_keep_structure(this->topology, HWLOC_OBJ_MACHINE) &&
+         /* ignore caches, which don't add structure */
+         0 == hwloc_topology_ignore_type_keep_structure(this->topology, HWLOC_OBJ_CORE) &&
+         0 == hwloc_topology_ignore_type_keep_structure(this->topology, HWLOC_OBJ_CACHE) &&
+         0 == hwloc_topology_set_flags(this->topology, HWLOC_TOPOLOGY_FLAG_WHOLE_SYSTEM) &&
+         #else
+         0 == hwloc_topology_set_all_types_filter(this->topology, HWLOC_TYPE_FILTER_KEEP_STRUCTURE) &&
+         #endif
+         0 == hwloc_topology_load(this->topology);
+   }
+#endif
+}
+
+void Machine_done(Machine* this) {
```

Comment:
Just curious. Did htop use the `_done` method name for a finalizer before? It's the first time I hear of this naming scheme.

Reply:
@Explorer09 yes, the pattern I've used here is based directly on the way ProcessList is handled (not too surprising  I guess as that's the structure/code being refactored here).

Reply:
Yes. This has been in the code base for ages. It's actually in nearly every (pseudo-)class. But the usage here is probably a bit inconsistent.

Trivia 1: There is the function pairs `_new` and `_delete` for memory self-allocation; while `_init`/`_done` is used if memory is provided.

Trivia 2: Actual Meter implementations do have an `_init` function but no `_new`. Instead for all meters a call to `Meter_new` is performed that is given the variable specifying the `MeterClass` to create. This makes all Meters effective sub-classes of the Meter class instead of true child classes created by inheritance.

Trivia 3: There's no `_new`/`_delete` for CRT, InfoScreen, and Platform.

Reply:
All platforms have been updated now - I've been able to verify Macos, Linux, PCP and 'unsupported' - if anyone with Solaris and *BSD machines handy could try a build on these, please let me know how it fares.  Thanks!